### PR TITLE
Add consistent object-based API for `EmailPasswordAuth` methods, deprecate positional argument API

### LINKED ISSUE #3943: Easy and consistent API for emailPassword methods
As reported in #3723, the API of `Realm.Auth.EmailPasswordAuth.resetPassword()` is not consistent between Realm Web and Realm JavaScript. Moreover, the TypeScript definition says on thing, and the API documentation says something else. The root cause is positional arguments (all strings) which is error prone for both our users and for ourself.

We should deprecate all current methods of `Realm.Auth.EmailPasswordAuth` (to be removed in version 11) and implement a new variant of the API:

```javascript
callResetPasswordFunction({ email: string, password: string }, args: [BSON])
confirmUser({ token: string, id: string })
registerUser({ email: string, password: string })
resendConfirmationEmail({ email: string }) // for consistency
resetPassword({ password: string, token: string, id: string })
sendResetPasswordEmail({ email: string }) // for consistency
```

Methods should log a warning if called with deprecated signature.

### PR DESCRIPTION
**Pull request by tomduncalf**
## What, How & Why?

This PR introduces a consistent, object/dictionary-based API for the `EmailPasswordAuth` methods, instead of using positional arguments, which can be error-prone and unclear. This closes #3943 and #3723.
 
The existing API is supported until the next major version, but will output deprecation warnings to the console when called in development mode of the form: `Deprecation warning from Realm: resetPassword(password, token, token_id) is deprecated and will be removed in a future major release. Consider switching to resetPassword({ password, token, token_id }).`

As the same logic is required to handle the deprecation in multiple places across `realm` and `realm-web`, and we may wish to deprecate more positional argument based APIs in future, a shared function `handleDeprecatedPositionalArgs` has been introduced, in a new Lerna package `@realm.io/common`. This function allows handling both the new and old forms of arguments in a consistent way, until the old way is deprecated at which point we can remove the call (see https://github.com/realm/realm-js/pull/4041#discussion_r733478780). This package can serve as a place to store other common code which does not warrant its own dedicated package between the two platforms in future.

### `registerUser`

Old:
```js
registerUser('john@example.com', 'some-secure-password');
```

New:
```js
registerUser({ email: 'john@example.com', password: 'some-secure-password' });
```

### `confirmUser`

Old:
```js
confirmUser('someToken', 'someTokenId');
```

New:
```js
confirmUser({ token: 'someToken', tokenId: 'someTokenId' })
```

### `resendConfirmationEmail`

Old:
```js
resendConfirmationEmail('john@example.com');
```

New:
```js
resendConfirmationEmail({ email: 'john@example.com' });
```

### `retryCustomConfirmation`

Old:
```js
retryCustomConfirmation('john@example.com');
```

New:
```js
retryCustomConfirmation({ email: 'john@example.com' });
```

### `sendResetPasswordEmail`

Old:
```js
sendPasswordEmail('john@example.com');
```

New:
```js
sendPasswordEmail({ email: 'john@example.com' });
```

### `resetPassword`

Old:
```js
resetPassword('some-secure-password', 'someResetToken', 'someResetTokenId');
```

New:
```js
resetPassword({ password: 'some-secure-password', token: 'someResetToken', tokenId 'someResetTokenId' });
```

### `callResetPasswordFunction`

Old:
```js
callResetPasswordFunction(
  'john@example.com', 
  'some-secure-password', 
  'extraArgToFunction1', 
  'extraArgToFunction2'
);
```

New:
```js
callResetPasswordFunction(
  { 
    email: 'john@example.com', 
    password: 'some-secure-password' 
  }, 
  'extraArgToFunction1', 
  'extraArgToFunction2'
);
```

## Questions:
- Do we want to only log warnings in development mode, or should we do it in all modes?
- Should we mention something in the Compatibility section?

## Notes:
- CI is failing until the `@realm.io/common` package is published, which I won't do until it has been reviewed. This is because currently our CI runs `npm install` before running Lerna, which pulls in code from npm only. I had a hacky(?) workaround in https://github.com/realm/realm-js/pull/4041/commits/c9d4bcef56090507a2ced962dab18539065b0a4b but after discussion with Kræn, we decided the real solution is to move `realm` to a Lerna package also, so for now we can accept having to first publish the package before CI passes.

## ☑️ ToDos
Todo:
* [x] Add workflow for running package tests
* [x] 📝 Changelog entry
* [x] 📝 `Compatibility` label is updated or copied from previous entry
* [x] 🚦 Tests
* [x] Test in sample apps etc
* [x] 📝 Public documentation PR created or is not necessary
  * https://github.com/mongodb/docs-realm/pull/1438
* [x] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [x] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: packages/realm-common/src/environment.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+/**
+ * Set the implementation used for `isDevelopmentMode`. This allows each entry point
+ * (node vs DOM) to specify an appropriate implementation, while maintaing a common
+ * interface.
+ *
+ * @param impl A function returning a boolean indicating whether the user's app is running
+ * in development mode or not.
+ */
+export const setIsDevelopmentModeImpl = (impl: typeof isDevelopmentModeImpl): void => {
```
Comment (kraenhansen) [inline packages/realm-common/src/environment.ts:39]: Is there a reason you choose a `callback` pattern over simply setting a boolean on import?

File: .github/workflows/package-unit-tests.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      - "src/**"
+      - "types/**"
+      - "react-native/**"
+      - "vendor/**"
+      # Integration tests
+      - "integration-tests/**"
+      # No need to run when updating documentation
+      - "!**.md"
+      # Run this on updates to the workflow
+      - ".github/workflows/integration-tests.yml"
```
Comment (kraenhansen) [inline .github/workflows/package-unit-tests.yml:22]: These paths looks like they we're copied 1-to-1.

```suggestion
      # Source code
      - "packages/realm-network-transport/**"
      - "packages/realm-common/**"
      # No need to run when updating documentation
      - "!**.md"
      # Run this on updates to the workflow
      - ".github/workflows/package-unit-tests.yml"
```

File: packages/realm-common/src/environment.ts (hunk 1 above)
Comment (tomduncalf) [reply packages/realm-common/src/environment.ts:39]: Hm, good point, the development mode status is not going to change in a given session. I'll simplify it

File: integration-tests/environments/react-native/package.json [hunk 3]
```diff
@@ -53,6 +53,8 @@
     "realm-app-importer": "*"
   },
   "localDependencies": {
+    "@realm.io/common": "../../../packages/realm-common",
+    "realm-network-transport": "../../../packages/realm-network-transport",
```
Comment (kraenhansen) [inline integration-tests/environments/react-native/package.json:57]: This shouldn't be needed (or they should be removed again) after https://github.com/realm/realm-js/pull/4054 merge (depending on what happens first 😅 )

Comment (tomduncalf) [reply integration-tests/environments/react-native/package.json:57]: 👍  I am just reviewing that one now so if we merge that first I can update this after!

File: packages/realm-common/src/environment.ts [hunk 4]
```diff
@@ -16,29 +16,26 @@
...
+export const isDevelopmentMode = (): boolean => isDevelopmentModeImpl();
 
-  if (typeof window !== "undefined") {
-    // We are in a browser, check if we are in react dev mode
-    return typeof global.__DEV__ !== "undefined" && global.__DEV__;
-  } else {
-    // We are not in a browser, check if NODE_ENV is anything other than production
-    return process.env?.NODE_ENV !== "production";
-  }
+/**
```
[the remark this answers is not in the document]
Comment (kraenhansen) [reply packages/realm-common/src/environment.ts:31]: Is there a reason you choose a callback pattern over simply setting a boolean on import?

**Pull request merged**